### PR TITLE
DQCP

### PR DESCRIPTION
**Pull request by akshayka**
Implementation of DQCP (http://web.stanford.edu/~boyd/papers/dqcp.html)

Adds an `is_dqcp()` method for expressions. problem.solve(qcp=True) will parse a problem as a DQCP and solve it with bisection. Adds a reduction Dqcp2Dcp that takes a DQCP problem and emits a parameterized DCP problem.

### CODE REVIEW

File: cvxpy/atoms/atom.py [hunk 1]
```diff
@@ -218,6 +229,57 @@ def is_log_log_concave(self):
         else:
             return False
 
+    @perf.compute_once
```
Comment (SteveDiamond) [inline cvxpy/atoms/atom.py]: How does this compute_once work? Can we use it more places?

File: cvxpy/atoms/dist_ratio.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        if not self.args[2].is_constant():
+            raise ValueError("`b` must be a constant.")
+        self.a = self.args[1].value
+        self.b = self.args[2].value
+
+    @Atom.numpy_numeric
+    def numeric(self, values):
+        """Returns the distance ratio.
+        """
+        return np.linalg.norm(
```
Comment (SteveDiamond) [inline cvxpy/atoms/dist_ratio.py]: This spacing looks kind of weird.

File: cvxpy/atoms/atom.py (hunk 1 above)
Comment (akshayka) [reply cvxpy/atoms/atom.py]: It's basically the same thing as lazy_prop, but not for properties. I've actually already sprinkled @perf.compute_once in various places at master: https://github.com/cvxgrp/cvxpy/search?utf8=%E2%9C%93&q=compute_once&type=

File: cvxpy/atoms/elementwise/ceil.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+import scipy.sparse as sp
+
+
+class ceil(Elementwise):
+    """Elementwise ceiling."""
+
+    def __init__(self, x):
+        return super(ceil, self).__init__(x)
+
+    @Elementwise.numpy_numeric
```
Comment (SteveDiamond) [inline cvxpy/atoms/elementwise/ceil.py]: Are we still doing numpy_numeric? I thought I had gotten rid of that.

File: cvxpy/reductions/dqcp2dcp/__init__.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
+"""
+Copyright 2017 Robin Verschueren
```
Comment (SteveDiamond) [inline cvxpy/reductions/dqcp2dcp/__init__.py]: Is this intended?

File: cvxpy/reductions/dqcp2dcp/dqcp2dcp.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    def _canonicalize_tree(self, expr):
+        canon_args, constrs = self._canon_args(expr)
+        canon_expr, c = self.canonicalize_expr(expr, canon_args)
+        constrs += c
+        return canon_expr, constrs
+
+    def _canon_args(self, expr):
+        """Canonicalize arguments of an expression.
+
+        Like Canonicalization.canonicalize_tree, but preserves signs.
```
Comment (SteveDiamond) [inline cvxpy/reductions/dqcp2dcp/dqcp2dcp.py]: Why doesn't canonicalize tree preserve signs? Aren't the variables the same?

File: cvxpy/reductions/solvers/bisection.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+def _find_bisection_interval(problem, t, solver=None, low=None, high=None):
+    """Finds an interval for bisection."""
+    if low is None:
+        low = 0 if t.is_nonneg() else -1
+    if high is None:
+        high = 0 if t.is_nonpos() else 1
+
+    infeasible_low = t.is_nonneg()
+    feasible_high = t.is_nonpos()
+    for _ in range(100):
```
Comment (SteveDiamond) [inline cvxpy/reductions/solvers/bisection.py]: Should this 100 be a parameter?

File: cvxpy/reductions/solvers/intermediate_chain.py [hunk 7]
```diff
@@ -49,13 +49,19 @@ def construct_intermediate_chain(problem, candidates, gp=False):
...
+            append = (" However, the problem does follow DQCP rules. "
+                      "Consider calling solve() with `qcp=True`.")
         raise DCPError("Problem does not follow DCP rules." + append)
     elif gp and not problem.is_dgp():
         append = ""
         if problem.is_dcp():
             append = (" However, the problem does follow DCP rules. "
-                      "Consider calling this function with `gp=False`.")
+                      "Consider calling solve() with `gp=False`.")
+        if problem.is_dqcp():
```
Comment (SteveDiamond) [inline cvxpy/reductions/solvers/intermediate_chain.py]: elif?

**Comment by SteveDiamond** [comment]
I did a quick look over and it seems pretty good. I made some random comments. I don't totally know what's happening in the code. You said you added lazy constraints. I would have imagined that the quasiconvex atoms would actually take a parameter when they were canonicalized, rather than making some kind of constraint and then canonicalizing that. Maybe I misunderstood your approach. It seems kind of indirect. 

What's an example of an atom that can't be canonicalized directly into a constraint with a parameter? Just trying to understand. Would it be like ``length(x) <= negative number``?

File: cvxpy/atoms/elementwise/ceil.py (hunk 3 above)
Comment (akshayka) [reply cvxpy/atoms/elementwise/ceil.py]: I copied it from some another atom's code. Deleted.

File: cvxpy/reductions/dqcp2dcp/__init__.py (hunk 4 above)
Comment (akshayka) [reply cvxpy/reductions/dqcp2dcp/__init__.py]: fixed, thanks

File: cvxpy/reductions/dqcp2dcp/dqcp2dcp.py (hunk 5 above)
Comment (akshayka) [reply cvxpy/reductions/dqcp2dcp/dqcp2dcp.py]: In regular canonicalization, when an expression like

nonlinear_atom(my_variable(pos=True))

is canonicalized, `nonlinear_atom` is replaced with a new slack variable, which doesn't have any attributes associated with it. This doesn't really matter for Dcp2Cone, Dgp2Dcp, or QP2QuadForm, but it does matter for Dqcp2Dcp. Should I move the logic in _canon_args into Canonicalization?

File: cvxpy/reductions/solvers/bisection.py (hunk 6 above)
Comment (akshayka) [reply cvxpy/reductions/solvers/bisection.py]: done

File: cvxpy/reductions/solvers/intermediate_chain.py (hunk 7 above)
Comment (akshayka) [reply cvxpy/reductions/solvers/intermediate_chain.py]: done, nice catch

File: cvxpy/atoms/dist_ratio.py (hunk 2 above)
Comment (akshayka) [reply cvxpy/atoms/dist_ratio.py]: I think it's standard? I can't think of a cleaner way to fit it that doesn't require `\` or wrapping the entire expression in (), both of which imo would make it harder to read.

**Comment by akshayka** [comment]
Yes, `length(arg) <= t` is a good example. You can take a look at the `length_sub` function in dqcp2dcp/sets.py, which constructs the sublevel set for length.

The natural way to write the t-sublevel set of len(arg) <= t is something like arg[floor(t):] == 0. But if t < 0, the constraint is infeasible (and can't be written in the aforementioned natural way), and if t > arg.size, then arg[floor(t):] is invalid syntax and will throw an error. Because t might be the bisection parameter, we don't know its value during canonicalization. When / if the bisection algorithm passes in t < 0 or t > arg.size in this example, we need a way to communicate that the constraint is infeasible or that the sublevel set is all of R^n, respectively. This lazy evaluation was the simplest way I could think of to handle this.

**Comment by akshayka** [comment]
Not sure what's going on with the windows build ... will take a look. @SteveDiamond , I'll let you know when the tests are passing.

File: cvxpy/atoms/dist_ratio.py (hunk 2 above)
Comment (SteveDiamond) [reply cvxpy/atoms/dist_ratio.py]: ok thanks, these are just random comments.

File: cvxpy/reductions/dqcp2dcp/dqcp2dcp.py (hunk 5 above)
Comment (SteveDiamond) [reply cvxpy/reductions/dqcp2dcp/dqcp2dcp.py]: No, don't change it. The variable attributes would produce constraints, which we don't want.

**Comment by SteveDiamond** [comment]
Ok sounds good. My guess would be something related to integer overflow (Windows has 32 bit integers for Python int, I think). I looked at the code more, and I understand your point better. @bstellato was actually suggesting lazy constraints for MIP problems, so it might be a concept worth introducing more broadly.

**Comment by akshayka** [comment]
Yes, it's probably under/overflow ... but in the solvers (I think). The problems that are failing have constraints like x >= inv_pos(y) * -t, y >= 0, and the LHS might be blowing up ... I'll try with ECOS instead of SCS and see what happens

**Comment by akshayka** [comment]
> @bstellato was actually suggesting lazy constraints for MIP problems, so it might be a concept worth introducing more broadly.

Yeah definitely! I don't yet have a good feeling for what the proper abstraction for lazy constraints would be, which is why I implemented them as a one-off thing for now. Once we collect more use cases, we can look at the requirements and design a good abstraction.

**Comment by akshayka** [comment]
@SteveDiamond , the tests are now passing.

**Comment by SteveDiamond** [comment]
Great! Merging in.

**Pull request merged**